contracts/core/EntryPoint.sol

```diff
@@ -219,29 +245,27 @@ contract EntryPoint is IEntryPoint, StakeManager {
 
     /**
      * Simulate a call to account.validateUserOp and paymaster.validatePaymasterUserOp.
-     * @dev this method always revert. Successful result is SimulationResult error. other errors are failures.
+     * @dev this method always revert. Successful result is ValidationResult error. other errors are failures.
      * @dev The node must also verify it doesn't use banned opcodes, and that it doesn't reference storage outside the account's data.
      * @param userOp the user operation to validate.
      */
-    function simulateValidation(UserOperation calldata userOp) external {
-        uint256 preGas = gasleft();
-
+    function validateUserOp(UserOperation calldata userOp) external {
         UserOpInfo memory outOpInfo;
 
-        (address aggregator, uint256 deadline) = _validatePrepayment(0, userOp, outOpInfo, SIMULATE_FIND_AGGREGATOR);
-        uint256 prefund = outOpInfo.prefund;
-        uint256 preOpGas = preGas - gasleft() + userOp.preVerificationGas;
+        (uint256 deadline, uint256 paymasterDeadline, address aggregator) = _validatePrepayment(0, userOp, outOpInfo, SIMULATE_FIND_AGGREGATOR);
         StakeInfo memory paymasterInfo = getStakeInfo(outOpInfo.mUserOp.paymaster);
         StakeInfo memory senderInfo = getStakeInfo(outOpInfo.mUserOp.sender);
         bytes calldata initCode = userOp.initCode;
         address factory = initCode.length >= 20 ? address(bytes20(initCode[0 : 20])) : address(0);
         StakeInfo memory factoryInfo = getStakeInfo(factory);
 
+        ReturnInfo memory returnInfo = ReturnInfo(outOpInfo.preOpGas, outOpInfo.prefund, deadline, paymasterDeadline, getMemoryBytesFromOffset(outOpInfo.contextOffset));
```

Comment:
Is there any explanation why `_validatePrepayment` returns `deadline`, `paymasterDeadline` and `aggregator` by value but fills in `preOpGas`, `prefund`, `contextOffset` etc. on a memory-pointer `UserOpInfo` struct?
Note that if there is no rational explanation that means that it is just a very very bad code, in my opinion.

Reply:
1. some of it is of course historic reasons.
2. parameters that are used directly are return values on the stack. parameters that are used later (sometimes even multiple times, in different contexts) or return multiple stack depths are in memory.

Reply:
Re #2, I looked at the code and I don't see that difference. The `aggregator` stack variable is usually not used (makes sense), and `outOpInfo` is immediately unwrapped into the `ReturnInfo` (this line).

I can't block the entire PR on this, but I would be less sad if the `_validatePrepayment` could return only an instance of `OutOpInfo` that is allocated inside the method.
Or, if that is expensive in gas, if it could pack all fields into the struct.

We may do these kinds of changes in a separate PR for a later revision of the EntryPoint.

Reply:
return a struct is inefficient in solidity - it requires in-memory copy. (it actually started that way, before I switched to an "out" reference).
please look at all calls to the methods.
UserOpInfo is the state that is kept per userOp between validation and execution. we need an array of those, since we perform all validations before all executions.
In simulateExecution (and in validateUserOp) we handle a single userOp, and thus we don't really need that, but we do want to use same inner functions.

Reply:
UserOpInfo contains exactly the parameters we later move into innerHandleOp, and thus we don't want to expand it (it requires an actual copy in to the inner context)
I did experiment with a larger struct, wrapping it, that contains parameters returned by the validation functions that are not required by innerHandleOp, instead of using return values.
it turned out it took even more gas - since stack is more efficient than memory (and as a side-effect this change also obfuscated the code even more and made it even less comprehensible, since with a referenced struct you can't tell which parameter is input and which is output)